### PR TITLE
Validation  Chapter for MAUI eBook

### PR DESCRIPTION
**Pull request by michaelstonis**
## Summary

Provides the validation chapter for the .NET MAUI eBook

### CODE REVIEW

File: docs/architecture/maui/validation.md [hunk 1]
```diff
@@ -1,9 +1,300 @@
...
+    public bool Check(T value)
+    {
+        if (value == null)
+        {
+            return false;
+        }
+
+        var str = value as string;
+        return !string.IsNullOrWhiteSpace(str);
+    }
```
Comment (IEvangelist) [inline docs/architecture/maui/validation.md:106]:
```suggestion
    public bool Check(T value) =>
        value is string str &&
        !string.IsNullOrWhiteSpace(str);
```

Algebraic reduction.

File: docs/architecture/maui/validation.md [hunk 2]
```diff
@@ -1,9 +1,300 @@
...
+        {
+            return false;
+        }
+
+        var str = value as string;
+        Regex regex = new Regex(@"^([w.-]+)@([w-]+)((.(w){2,3})+)$");
+        Match match = regex.Match(str);
+
+        return match.Success;
+    }
```
Comment (IEvangelist) [inline docs/architecture/maui/validation.md:131]:
```suggestion
    private readonly Regex _regex = new(@"^([w.-]+)@([w-]+)((.(w){2,3})+)$");

    public bool Check(T value) =>
        value is string str && _regex.IsMatch(str);
```

Algebraic reduction, and avoid cache regex in a class-scoped field.

File: docs/architecture/maui/validation.md [hunk 3]
```diff
@@ -1,9 +1,300 @@
...
+}
+```
+
+The `Check` method returns a boolean indicating whether or not the value argument is a valid email address. This is achieved by searching the value argument for the first occurrence of the regular expression pattern specified in the Regex constructor. Whether the regular expression pattern has been found in the input string can be determined by checking the value of the Match object's Success property.
```
Comment (IEvangelist) [inline docs/architecture/maui/validation.md:135]:
```suggestion
The `Check` method returns a boolean indicating whether or not the value argument is a valid email address. This is achieved by searching the value argument for the first occurrence of the regular expression pattern specified in the `Regex` constructor. Whether the regular expression pattern has been found in the input string can be determined by checking the `value` against <xref:System.Text.RegularExpressions.Regex.IsMatch%2A?displayProperty=nameWithType>.
```

File: docs/architecture/maui/validation.md [hunk 4]
```diff
@@ -1,9 +1,300 @@
...
+
+public ValidatableObject<string> UserName
+{
+    get => _userName;
+}
+
+public ValidatableObject<string> Password
+{
+    get => _password;
+}
```
Comment (IEvangelist) [inline docs/architecture/maui/validation.md:156]:
```suggestion
  public ValidatableObject<string> UserName { get; private set; }

  public ValidatableObject<string> Password { get; private set; }
```

Use auto-properties when possible.

**Review by IEvangelist — CHANGES_REQUESTED** [review]
Looks great, I'm curious, how much of this is homegrown versus built-in? It looks like the `ValidatableObject<T>` is custom. Does .NET MAUI not yet support DataAnnotations for validation?

File: docs/architecture/maui/validation.md [hunk 5]
```diff
@@ -1,9 +1,300 @@
...
+description: Provide feedback and checks to ensure data is correct at runtime
 author: michaelstonis
 no-loc: [MAUI]
-ms.date: 07/12/2022
+ms.date: 07/09/2022
 ---
 
 # Validation
+
+Any app that accepts input from users should ensure that the input is valid. An app could, for example, check for input that contains only characters in a particular range, is of a certain length, or matches a particular format. Without validation, a user can supply data that causes the app to fail. Validation enforces business rules, and prevents an attacker from injecting malicious data.
```
[the remark this answers is not in the document]
Comment (michaelstonis) [reply docs/architecture/maui/validation.md:11]: I like it and agree.

**Comment by michaelstonis** [comment]
> Looks great, I'm curious, how much of this is homegrown versus built-in? It looks like the `ValidatableObject<T>` is custom. Does .NET MAUI not yet support DataAnnotations for validation?

Unfortunately we don't have any built-in support for data annotations and similar validation constructs. For mobile we either need to use a homegrown solution or a 3rd party component (e.g. [ObservableValidator](https://docs.microsoft.com/en-us/windows/communitytoolkit/mvvm/observablevalidator)).

**Review by IEvangelist — APPROVED** [review]

**Pull request merged**